src/renderer.cpp

```diff
@@ -1357,6 +1604,248 @@ void Renderer::setD3DProjection(struct matrix* matrix)
     if (uniform_log) printMatrix(__func__, internalState.d3dProjectionMatrix);
 };
 
+void Renderer::setWorldLightDir(float dirX, float dirY, float dirZ)
+{
+    internalState.worldLightRot.x = dirX;
+    internalState.worldLightRot.y = dirY;
+    internalState.worldLightRot.z = dirZ;
+}
+
+struct point3d Renderer::getWorldLightDir()
+{
+    return internalState.worldLightRot;
+}
+
+void Renderer::setLightIntensity(float intensity)
+{
+    internalState.lightData[3] = intensity;
+}
+
+float Renderer::getLightIntensity()
+{
+    return internalState.lightData[3];
+}
+
+void Renderer::setLightColor(float r, float g, float b)
+{
+    internalState.lightData[0] = r;
+    internalState.lightData[1] = g;
+    internalState.lightData[2] = b;
+}
+
+point3d Renderer::getLightColor()
+{
+    struct point3d color = { internalState.lightData[0],
+                             internalState.lightData[1],
+                             internalState.lightData[2] };
+    return color;
+}
+
+void Renderer::setAmbientIntensity(float intensity)
+{
+    internalState.ambientLightData[3] = intensity;
+}
+
+float Renderer::getAmbientIntensity()
+{
+    return internalState.ambientLightData[3];
+}
+
+void Renderer::setAmbientLightColor(float r, float g, float b)
+{
+    internalState.ambientLightData[0] = r;
+    internalState.ambientLightData[1] = g;
+    internalState.ambientLightData[2] = b;
+}
+
+point3d Renderer::getAmbientLightColor()
+{
+    struct point3d color = { internalState.ambientLightData[0],
+                             internalState.ambientLightData[1],
+                             internalState.ambientLightData[2] };
+    return color;
+}
+
+void Renderer::setRoughness(float roughness)
+{
+    internalState.materialData[0] = roughness;
+}
+
+float Renderer::getRoughness()
+{
+    return internalState.materialData[0];
+}
+
+void Renderer::setMetalness(float metalness)
+{
+    internalState.materialData[1] = metalness;
+}
+
+float Renderer::getMetalness()
+{
+    return internalState.materialData[1];
+}
+
+void Renderer::setShadowFaceCullingEnabled(bool isEnabled)
+{
+    internalState.isShadowMapFaceCullingEnabled = isEnabled;
+}
+
+bool Renderer::isShadowFaceCullingEnabled()
+{
+    return internalState.isShadowMapFaceCullingEnabled;
+}
+
+void Renderer::setShadowMapResolution(int resolution)
+{
+    internalState.shadowData[3] = resolution;
+
+    if (bgfx::isValid(shadowMapFrameBuffer))
+    {
+        bgfx::destroy(shadowMapFrameBuffer);
+    }
+
+    shadowMapTexture = bgfx::createTexture2D(
+        resolution,
+        resolution,
+        false,
+        1,
+        bgfx::TextureFormat::D32,
+        BGFX_TEXTURE_RT | BGFX_SAMPLER_COMPARE_LEQUAL | BGFX_SAMPLER_U_CLAMP | BGFX_SAMPLER_V_CLAMP | BGFX_SAMPLER_W_CLAMP
+    );
+
+    shadowMapFrameBuffer = bgfx::createFrameBuffer(
+        1,
+        &shadowMapTexture,
+        true
+    );

```

Comment:
Please consider reworking this entire code to recreate the framebuffer on the `reset()` method call only.

Reply:
If I understand correctly, implementing it inside `reset()` would mean that shadow map resolution can only be set during initialization or when the window size is resized. However, shadow map resolution can also be set from the lighting debug window so moving this code to the `reset()` would not work.

Reply:
Debug code should have its own special workflow. If it needs to reset the entire view, so be it. It's a dev tool afterall.

Generically speaking we develop towards optimizations for the generic experience, and because of this, I rather prefer this code being in the `reset()`. Debug code shall never degrade the overall code quality, just because it's there.